src/optimizer/query_to_operator_transformer.cpp

```diff
@@ -272,21 +263,139 @@ void QueryToOperatorTransformer::Visit(
+      expr->SetChild(1, select_list.at(0)->Copy());
+      expr->SetExpressionType(ExpressionType::COMPARE_EQUAL);
+    }
+
+  } else if (expr_type == ExpressionType::COMPARE_EQUAL ||
+             expr_type == ExpressionType::COMPARE_GREATERTHAN ||
+             expr_type == ExpressionType::COMPARE_GREATERTHANOREQUALTO ||
+             expr_type == ExpressionType::COMPARE_LESSTHAN ||
+             expr_type == ExpressionType::COMPARE_LESSTHANOREQUALTO) {
+    std::vector<expression::AbstractExpression *> select_list;
+    if (GenerateSubquerytree(expr->GetModifiableChild(0), select_list, true)) {
+      if (select_list.size() != 1)
+        throw Exception("Array in predicates not supported");
+
+      // Set the left child as the output of the subquery
+      expr->SetChild(0, select_list.at(0)->Copy());
+    }
+    select_list.clear();
+    if (GenerateSubquerytree(expr->GetModifiableChild(1), select_list, true)) {
+      if (select_list.size() != 1)
+        throw Exception("Array in predicates not supported");
+
+      // Set the right child as the output of the subquery
+      expr->SetChild(1, select_list.at(0)->Copy());
+    }
+  }
+  expr->AcceptChildren(this);
+}
+
+void QueryToOperatorTransformer::Visit(expression::OperatorExpression *expr) {
+  if (expr->GetExpressionType() == ExpressionType::OPERATOR_EXISTS) {
+    std::vector<expression::AbstractExpression *> select_list;
+    if (GenerateSubquerytree(expr->GetModifiableChild(0), select_list)) {
+      PL_ASSERT(!select_list.empty());
+
+      // Set the right child as the output of the subquery
+      expr->SetChild(0, select_list.at(0)->Copy());
+    }
+  }
+
+  expr->AcceptChildren(this);
+}
+
+bool QueryToOperatorTransformer::RequireAggregation(
+    const parser::SelectStatement *op) {
+  if (op->group_by != nullptr) return true;
+  // Check plain aggregation
+  bool has_aggregation = false;
+  bool has_other_exprs = false;
+  for (auto &expr : op->getSelectList()) {
+    std::vector<expression::AggregateExpression *> aggr_exprs;
+    expression::ExpressionUtil::GetAggregateExprs(aggr_exprs, expr.get());
+    if (!aggr_exprs.empty())
+      has_aggregation = true;
+    else
+      has_other_exprs = true;
+  }
+  // TODO: Should be handled in the binder
+  // Syntax error when there are mixture of aggregation and other exprs
+  // when group by is absent
+  if (has_aggregation && has_other_exprs)
+    throw SyntaxException(
+        "Non aggregation expression must appear in the GROUP BY "
+        "clause or be used in an aggregate function");
+  return has_aggregation;
+}
+
+void QueryToOperatorTransformer::CollectPredicates(
+    expression::AbstractExpression *expr) {
+  expr->Accept(this);
+  util::ExtractPredicates(expr, predicates_);
+}
+
+bool QueryToOperatorTransformer::GenerateSubquerytree(
+    expression::AbstractExpression *expr,
+    std::vector<expression::AbstractExpression *> &select_list,
+    bool single_join) {
```

Comment:
I don't understand this... what are we doing here? What are the input arguments? Why we're also handling joins here?

Reply:
Single join is a specific join semantic that only produce the join result if the right relation has less than or equal to 1 tuple for each tuple from the left relation, or else it would raise an exception. We perform unnesting by transforming a select in an expression to a new join operator including dependent join, single join and mark join. Detail see [This paper] (http://btw2017.informatik.uni-stuttgart.de/slidesandpapers/F1-10-37/paper_web.pdf). We'll add some comments for these new operators.

Reply:
I see.